# **Uplift Modelling of a Marketing Campaign.**

## **Load Data.**

In [1]:
import os
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

DATASETS_PATH = os.getenv("DATASETS_PATH")
DATASETS_PATH = DATASETS_PATH + "/MarketingUpliftModelling/"

data_df = pd.read_csv(DATASETS_PATH + "Kevin_Hillstrom_MineThatData_E-MailAnalytics.csv")

print("Data: ")
print(data_df)

Data: 
       recency history_segment  history  mens  womens   zip_code  newbie  \
0           10  2) $100 - $200   142.44     1       0  Surburban       0   
1            6  3) $200 - $350   329.08     1       1      Rural       1   
2            7  2) $100 - $200   180.65     0       1  Surburban       1   
3            9  5) $500 - $750   675.83     1       0      Rural       1   
4            2    1) $0 - $100    45.34     1       0      Urban       0   
...        ...             ...      ...   ...     ...        ...     ...   
63995       10  2) $100 - $200   105.54     1       0      Urban       0   
63996        5    1) $0 - $100    38.91     0       1      Urban       1   
63997        6    1) $0 - $100    29.99     1       0      Urban       1   
63998        1  5) $500 - $750   552.94     1       0  Surburban       1   
63999        1  4) $350 - $500   472.82     0       1  Surburban       0   

            channel        segment  visit  conversion  spend  
0             Pho

## **Encode categorical string variables.**

In [2]:
df_encoded = pd.get_dummies(data_df, columns=["zip_code", "channel"], drop_first = True)

df_encoded["history_segment_num"] = df_encoded["history_segment"].str.extract(r"(\d+)\)").astype(int)

print(df_encoded)

       recency history_segment  history  mens  womens  newbie        segment  \
0           10  2) $100 - $200   142.44     1       0       0  Womens E-Mail   
1            6  3) $200 - $350   329.08     1       1       1      No E-Mail   
2            7  2) $100 - $200   180.65     0       1       1  Womens E-Mail   
3            9  5) $500 - $750   675.83     1       0       1    Mens E-Mail   
4            2    1) $0 - $100    45.34     1       0       0  Womens E-Mail   
...        ...             ...      ...   ...     ...     ...            ...   
63995       10  2) $100 - $200   105.54     1       0       0    Mens E-Mail   
63996        5    1) $0 - $100    38.91     0       1       1    Mens E-Mail   
63997        6    1) $0 - $100    29.99     1       0       1    Mens E-Mail   
63998        1  5) $500 - $750   552.94     1       0       1  Womens E-Mail   
63999        1  4) $350 - $500   472.82     0       1       0    Mens E-Mail   

       visit  conversion  spend  zip_co

## **Encode treatment variable and create features and target frames.**

In [3]:
df_encoded["treatment"] = (df_encoded["segment"] != "No E-Mail").astype(int)

df_encoded["recency_x_treatment"] = df_encoded["recency"] * df_encoded["treatment"]
df_encoded["history_x_treatment"] = df_encoded["history"] * df_encoded["treatment"]
df_encoded["newbie_x_treatment"] = df_encoded["newbie"] * df_encoded["treatment"]

df_encoded["history_per_recency"] = df_encoded["history"] / (df_encoded["recency"] + 1)
df_encoded["log_history"] = np.log1p(df_encoded["history"])

y = df_encoded["conversion"]
treatment = df_encoded["treatment"]

X = df_encoded.drop(columns = ["history_segment", "segment", "visit", "conversion", "spend", "treatment"])

print(X)

       recency  history  mens  womens  newbie  zip_code_Surburban  \
0           10   142.44     1       0       0                True   
1            6   329.08     1       1       1               False   
2            7   180.65     0       1       1                True   
3            9   675.83     1       0       1               False   
4            2    45.34     1       0       0               False   
...        ...      ...   ...     ...     ...                 ...   
63995       10   105.54     1       0       0               False   
63996        5    38.91     0       1       1               False   
63997        6    29.99     1       0       1               False   
63998        1   552.94     1       0       1                True   
63999        1   472.82     0       1       0                True   

       zip_code_Urban  channel_Phone  channel_Web  history_segment_num  \
0               False           True        False                    2   
1               False  

Here, the column 'history_segment' is encoded with integer numbers, because this category has an obvious order. 

## **Train an ensamble uplift model consisting of lightGBM.**

Uplift modeling requires estimating the difference between two conditional probabilities (treatment vs control). This makes the task inherently noisy and high-variance, especially in marketing datasets where treatment effects are small. Initial experiments with single uplift models (S-learner and tree-based approaches) showed unstable performance across different train-test splits, with Qini scores varying significantly. This indicates high variance in treatment effect estimation. To improve robustness, an ensemble approach was used. By averaging predictions from multiple models trained with different random seeds, variance in uplift estimation is reduced, leading to more stable and reliable customer ranking.

In [4]:
from lightgbm import LGBMClassifier
import numpy as np
from sklearn.model_selection import train_test_split
from sklift.metrics import qini_auc_score

X_train, X_test, y_train, y_test, t_train, t_test = train_test_split(
    X, y, treatment, test_size=0.3, random_state=42, stratify=treatment
)

n_models = 20
models = []

for seed in range(n_models):

    model = LGBMClassifier(
        n_estimators=110,
        learning_rate=0.04,
        num_leaves=4,
        max_depth=2,
        subsample=0.1,
        colsample_bytree=0.8,
        random_state=seed
    )

    X_train_s = X_train.copy()
    X_train_s["treatment"] = t_train

    model.fit(X_train_s, y_train)
    models.append(model)

def predict_uplift(model, X):
    X_t = X.copy()
    X_c = X.copy()

    X_t["treatment"] = 1
    X_c["treatment"] = 0

    return (
        model.predict_proba(X_t)[:, 1]
        - model.predict_proba(X_c)[:, 1]
    )

uplift_matrix = np.array([
    predict_uplift(m, X_test) for m in models
])

uplift_test = uplift_matrix.mean(axis=0)

from sklift.metrics import qini_auc_score

qini = qini_auc_score(
    y_true=y_test,
    uplift=uplift_test,
    treatment=t_test
)
print()
print()
print("Ensemble Qini:", qini)


[LightGBM] [Info] Number of positive: 397, number of negative: 44403
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000461 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1072
[LightGBM] [Info] Number of data points in the train set: 44800, number of used features: 16
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.008862 -> initscore=-4.717126
[LightGBM] [Info] Start training from score -4.717126
[LightGBM] [Info] Number of positive: 397, number of negative: 44403
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000403 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1072
[LightGBM] [Info] Number of data points in the train set: 44800, number of used features: 16
[LightGBM] [Info] [bin

In [5]:
uplift_full = np.mean([
    predict_uplift(m, X) for m in models
], axis=0)

data_df["uplift_score"] = uplift_full
print(data_df)

       recency history_segment  history  mens  womens   zip_code  newbie  \
0           10  2) $100 - $200   142.44     1       0  Surburban       0   
1            6  3) $200 - $350   329.08     1       1      Rural       1   
2            7  2) $100 - $200   180.65     0       1  Surburban       1   
3            9  5) $500 - $750   675.83     1       0      Rural       1   
4            2    1) $0 - $100    45.34     1       0      Urban       0   
...        ...             ...      ...   ...     ...        ...     ...   
63995       10  2) $100 - $200   105.54     1       0      Urban       0   
63996        5    1) $0 - $100    38.91     0       1      Urban       1   
63997        6    1) $0 - $100    29.99     1       0      Urban       1   
63998        1  5) $500 - $750   552.94     1       0  Surburban       1   
63999        1  4) $350 - $500   472.82     0       1  Surburban       0   

            channel        segment  visit  conversion  spend  uplift_score  
0         

The ansamble **Qini-Score** in this case is around 0.22, which is an excelent performance for an uplift model. So no further model fine-tuning is needed.

## **Compare the trained ensamble uplift models to a random selection.**

In [6]:
ks = [0.05, 0.1, 0.2]

data_df_sorted = data_df.sort_values("uplift_score", ascending=False)

for k in ks:
    n_target = int(len(data_df) * k)

    uplift_target = data_df_sorted.head(n_target)
    random_target = data_df.sample(n=n_target, random_state=42)

    uplift_rate = uplift_target["conversion"].mean()
    random_rate = random_target["conversion"].mean()

    absolute_lift = uplift_rate - random_rate
    relative_lift = uplift_rate / random_rate if random_rate > 0 else np.nan

    print(f"\n--- Top {int(k*100)}% targeting ---")
    print("Uplift policy response rate:", uplift_rate)
    print("Random policy response rate:", random_rate)
    print("Absolute uplift gain:", absolute_lift)
    print("Relative lift:", relative_lift)
    


--- Top 5% targeting ---
Uplift policy response rate: 0.01875
Random policy response rate: 0.0078125
Absolute uplift gain: 0.0109375
Relative lift: 2.4

--- Top 10% targeting ---
Uplift policy response rate: 0.01625
Random policy response rate: 0.00890625
Absolute uplift gain: 0.007343750000000001
Relative lift: 1.824561403508772

--- Top 20% targeting ---
Uplift policy response rate: 0.014921875
Random policy response rate: 0.01
Absolute uplift gain: 0.004921874999999999
Relative lift: 1.4921875


Considering **top 5% targeting**, an absolute uplift gain of around 1.09 percentage points means that for every 100 customers, random targeting delivers approximately 0.78 conversions, while the uplift model delivers about 1.88 conversions. This corresponds to approximately 1.1 additional conversions per 100 customers targeted compared to random selection. Overall, the uplift-based **targeting strategy is about 2.4 times more effective than random targeting** in terms of conversion rate.

Drastic decrease in relative lift for top 10% and 20% is due to the very good performance of the model which is highly effective at identifying a small group of strong persuadables, but the predicted uplift quality degrades as one includes lower-ranked customers.

## **Define the customers into 'persuadables', 'sure thing' and 'do not disturb' groups.**

In [7]:
X_t = X.copy()
X_c = X.copy()

X_t["treatment"] = 1
X_c["treatment"] = 0

p_t = np.mean([m.predict_proba(X_t)[:, 1] for m in models], axis=0)
p_c = np.mean([m.predict_proba(X_c)[:, 1] for m in models], axis=0)

data_df["p_t"] = p_t
data_df["p_c"] = p_c

uplift_thr = data_df["uplift_score"].quantile(0.8)
neg_thr = data_df["uplift_score"].quantile(0.2)
pc_thr = data_df["p_c"].median()

def segment_customer(row):
    if row["uplift_score"] > uplift_thr:
        return "Persuadable"

    elif row["uplift_score"] < neg_thr:
        return "Do Not Disturb"

    elif row["p_c"] > pc_thr:
        return "Sure Thing"

    else:
        return "Lost Cause"

data_df["uplift_segment"] = data_df.apply(segment_customer, axis=1)
print(data_df[data_df["uplift_segment"] == "Persuadable"])

       recency history_segment  history  mens  womens   zip_code  newbie  \
2            7  2) $100 - $200   180.65     0       1  Surburban       1   
3            9  5) $500 - $750   675.83     1       0      Rural       1   
4            2    1) $0 - $100    45.34     1       0      Urban       0   
6            9  3) $200 - $350   280.20     1       0  Surburban       1   
9           10    1) $0 - $100    32.84     0       1      Urban       1   
...        ...             ...      ...   ...     ...        ...     ...   
63968       11  2) $100 - $200   195.12     0       1  Surburban       1   
63969        3    1) $0 - $100    67.78     0       1  Surburban       0   
63986        9    1) $0 - $100    35.26     0       1      Urban       1   
63989       10  3) $200 - $350   304.30     1       1  Surburban       0   
63998        1  5) $500 - $750   552.94     1       0  Surburban       1   

            channel        segment  visit  conversion  spend  uplift_score  \
2        

## **Create pursuadables profile considering features of the customers.**

In [8]:
persuadables = data_df[data_df["uplift_segment"] == "Persuadable"]
overall = data_df.copy()

features_num = ["recency", "history"]

print("Persuadables:")
print(persuadables[features_num].mean())

print("\nOverall:")
print(overall[features_num].mean())

features_cat = ["channel", "zip_code", "newbie", "mens", "womens"]

for col in features_cat:
    print(f"\n{col} distribution (Persuadables):")
    print(persuadables[col].value_counts(normalize=True))
    
    print(f"\n{col} distribution (Overall):")
    print(overall[col].value_counts(normalize=True))

def compare_distribution(df1, df2, col):
    dist1 = df1[col].value_counts(normalize=True)
    dist2 = df2[col].value_counts(normalize=True)
    
    return (dist1 - dist2).sort_values(ascending=False)

compare_distribution(persuadables, overall, "channel")

importance = models[0].feature_importances_
feat_names = X.columns.tolist() + ["treatment"]

feature_importance = pd.Series(importance, index=feat_names).sort_values(ascending=False)
print()
print("Feature Importance: ")
print(str(feature_importance))

Persuadables:
recency      6.885857
history    320.410435
dtype: float64

Overall:
recency      5.763734
history    242.085656
dtype: float64

channel distribution (Persuadables):
channel
Web             0.626456
Multichannel    0.197326
Phone           0.176218
Name: proportion, dtype: float64

channel distribution (Overall):
channel
Web             0.440891
Phone           0.437828
Multichannel    0.121281
Name: proportion, dtype: float64

zip_code distribution (Persuadables):
zip_code
Surburban    0.449457
Urban        0.402236
Rural        0.148307
Name: proportion, dtype: float64

zip_code distribution (Overall):
zip_code
Surburban    0.449625
Urban        0.400953
Rural        0.149422
Name: proportion, dtype: float64

newbie distribution (Persuadables):
newbie
1    0.625909
0    0.374091
Name: proportion, dtype: float64

newbie distribution (Overall):
newbie
1    0.50225
0    0.49775
Name: proportion, dtype: float64

mens distribution (Persuadables):
mens
1    0.518333
0    0.48

## **Core profile of the pursuadables.**

Comparative analysis of persuadable vs overall customers
* **More valuable customers:** 'history' is 320 vs 242 (around 1.32 times higher), indicating persuadable customers tend to have significantly higher historical spending.
* **Slightly less recent activity:** 'recency' is 6.89 vs 5.76, meaning persuadable customers are actually slightly less recent buyers on average, suggesting that responsiveness is not purely driven by recency.
* **Strong new customer signal:** 'newbie' rate is 62.6% vs 50.2%, showing that new customers are substantially more likely to be persuadable, making lifecycle stage an important driver of uplift.
* **Multichannel interaction strongly overrepresented:** 19.7% vs 12.1% (around 1.6 times higher), indicating that multichannel customers are more responsive to campaigns and more likely to benefit from targeting.
* **Phone channel underrepresented:** 17.6% vs 43.8%, suggesting that phone-heavy users are significantly less responsive to email-based interventions.
* **Web channel overrepresented:** 62.6% vs 44.1%, showing that web-engaged users are more likely to be persuadable.
* **Women’s category strongly overrepresented:** 62.7% vs 54.9%, indicating a meaningful uplift responsiveness among women-focused customers.
* **Men’s category slightly underrepresented:** 51.8% vs 55.1%, suggesting weaker uplift signal for male-targeted segments.

Geography shows minimal difference:<br/>

Suburban: 44.9% vs 44.9%<br/>
Urban: 40.2% vs 40.1%<br/>
Rural: 14.8% vs 14.9%<br/>

This confirms that geographic variables are not meaningful drivers of uplift in this dataset.

**Feature importance analysis confirms this profile**<br/>

The model is primarily driven by:

* **history_per_recency** (dominant feature) - strongest predictor of responsiveness and engagement intensity
* **history_x_treatment** - key interaction capturing heterogeneous treatment effect by customer value
* **newbie** - strong lifecycle sensitivity signal
* **recency_x_treatment** - treatment effectiveness varies with engagement timing
* **history & recency** - baseline behavioral strength

Secondary contributors include:<br/>
* **channel effects (Phone, Web)**<br/>
* **product affinity (womens)**<br/>
* **log-transformed spending**<br/>

## **Final conclusion about this marketing campaign.**
Persuadable customers are primarily higher-value, moderately engaged users with strong interaction effects between historical spending and treatment exposure. They are disproportionately new customers, more likely to engage via web and multichannel touchpoints, and show stronger affinity for women-related products. Geographic factors contribute negligibly to uplift segmentation, while behavioral and interaction-based features dominate model decisions.